File: app/src/main/java/dev/yiray/qickmvivm/repository/UserRepository.java
package dev.yiray.qickmvivm.repository;

import dev.yiray.qickmvivm.model.User;
import dev.yiray.qickmvivm.retrofit.UserInterface;
import dev.yiray.qickmvivm.retrofit.UserService;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class UserRepository {
    private final UserInterface userInterface;
    private static UserRepository repository;

    public interface OnInfoCallBack {
        void onInformationTaken(User user);
    }

    public static UserRepository getInstance(){
        if (repository == null){
            repository = new UserRepository();
        }
        return repository;
    }

    public UserRepository(){
        userInterface = UserService.getInterface();
    }

    public void getUser(String userName, OnInfoCallBack callBack) {
        userInterface.findUser(userName).enqueue(new Callback<User>() {
            @Override
            public void onResponse(Call<User> call, Response<User> response) {
                callBack.onInformationTaken(response.body());
            }

            @Override
            public void onFailure(Call<User> call, Throwable t) {
                callBack.onInformationTaken(null);
            }
        });
    }

}

File: app/src/main/java/dev/yiray/livedatamvi/repository/UserRepository.kt
package dev.yiray.livedatamvi.repository

import dev.yiray.livedatamvi.retrofit.UserService
import dev.yiray.livedatamvi.retrofit.UserApi
import dev.yiray.livedatamvi.model.User
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class UserRepository {
    private val userApi: UserApi?

    interface OnInfoCallBack {
        fun onInformationTaken(user: User?)
    }

    fun getUser(userName: String?, callBack: OnInfoCallBack) {
        userApi!!.findUser(userName).enqueue(object : Callback<User?> {
            override fun onResponse(call: Call<User?>, response: Response<User?>) {
                callBack.onInformationTaken(response.body())
            }

            override fun onFailure(call: Call<User?>, t: Throwable) {
                callBack.onInformationTaken(null)
            }
        })
    }

    companion object {
        private var repository: UserRepository? = null
        val instance: UserRepository?
            get() {
                if (repository == null) {
                    repository = UserRepository()
                }
                return repository
            }
    }

    init {
        userApi = UserService.api
    }
}